apiserver/facades/client/resources/repository.go

```diff
@@ -0,0 +1,279 @@
+
+	// ResourceInfo returns the metadata for the given resource.
+	ResourceInfo(charmstore.ResourceRequest) (charmresource.Resource, error)
+}
+
+type charmStoreClient struct {
+	client CharmStore
+	id     CharmID
+}
+
+func (cs *charmStoreClient) ResolveResources(resources []charmresource.Resource) ([]charmresource.Resource, error) {
+	storeResources, err := cs.resourcesFromCharmstore()
+	if err != nil {
+		return nil, err
+	}
+	resolved, err := cs.resolveResources(resources, storeResources)
+	if err != nil {
+		return nil, err
+	}
+	// TODO(ericsnow) Ensure that the non-upload resource revisions
+	// match a previously published revision set?
+	return resolved, nil
+}
+
+// resourcesFromCharmstore gets the info for the charm's resources in
+// the charm backend. If the charm URL has a revision then that revision's
+// resources are returned. Otherwise the latest info for each of the
+// resources is returned.
+func (cs *charmStoreClient) resourcesFromCharmstore() (map[string]charmresource.Resource, error) {
+	results, err := cs.listResources([]CharmID{cs.id})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	storeResources := make(map[string]charmresource.Resource)
+	if len(results) != 0 {
+		for _, res := range results[0] {
+			storeResources[res.Name] = res
+		}
+	}
+	return storeResources, nil
+}
+
+// resolveResources determines the resource info that should actually
+// be stored on the controller. That decision is based on the provided
+// resources along with those in the charm backend (if any).
+func (cs *charmStoreClient) resolveResources(
+	resources []charmresource.Resource,
+	storeResources map[string]charmresource.Resource,
+) ([]charmresource.Resource, error) {
+	allResolved := make([]charmresource.Resource, len(resources))
+	copy(allResolved, resources)
+	for i, res := range resources {
+		// Note that incoming "upload" resources take precedence over
+		// ones already known to the controller, regardless of their
+		// origin.
+		if res.Origin != charmresource.OriginStore {
+			continue
+		}
+
+		resolved, err := cs.resolveStoreResource(res, storeResources)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		allResolved[i] = resolved
+	}
+	return allResolved, nil
+}
+
+// resolveStoreResource selects the resource info to use. It decides
+// between the provided and latest info based on the revision.
+func (cs *charmStoreClient) resolveStoreResource(
+	res charmresource.Resource,
+	storeResources map[string]charmresource.Resource,
+) (charmresource.Resource, error) {
+	storeRes, ok := storeResources[res.Name]
+	if !ok {
+		// This indicates that AddPendingResources() was called for
+		// a resource the charm backend doesn't know about (for the
+		// relevant charm revision).
+		// TODO(ericsnow) Do the following once the charm backend supports
+		// the necessary endpoints:
+		// return res, errors.NotFoundf("charm backend resource %q", res.Name)
+		return res, nil
```

Comment:
We should can this comment. We should make a bug instead.

Reply:
https://bugs.launchpad.net/juju/+bug/1906838